overlord/snapstate/handlers.go

```diff
@@ -310,20 +334,120 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
 	return ts, err
 }
 
-func (m *SnapManager) installPrereqs(t *state.Task, base string, prereq []string, userID int, tm timings.Measurer) error {
+// updates a prerequisite, if it's not providing a content interface that a plug expects it to
+func updatePrereqIfOutdated(t *state.Task, snapName string, contentAttrs []string, userID int, flags Flags) (*state.TaskSet, error) {
+	st := t.State()
+
+	if len(contentAttrs) == 0 {
+		return nil, nil
+	}
+
+	// check if the default provider has all expected content tags
+	if ok, err := hasAllContentAttrs(st, snapName, contentAttrs); err != nil {
+		return nil, err
+	} else if ok {
+		return nil, nil
+	}
+
+	if ok, err := shouldSkipToAvoidConflict(t, snapName); err != nil {
+		return nil, err
+	} else if ok {
+		return nil, nil
+	}
+
+	// default provider is missing some content tags (likely outdated) so update it
+	ts, err := Update(st, snapName, nil, userID, flags)
+	if err != nil {
+		if conflErr, ok := err.(*ChangeConflictError); ok {
+			// there's already an update for the same snap in this change,
+			// just skip this one
+			if conflErr.ChangeID == t.Change().ID() {
+				return nil, nil
+			}
+
+			return nil, &state.Retry{After: prerequisitesRetryTimeout}
+		}
+
+		// don't propagate error to avoid failing the main install since the
+		// content provider is (for now) a soft dependency
+		t.Logf("installed prerequisite '%s' is missing content. Failed to update: %s", snapName, err)
```

Comment:
I wonder if we should maybe reorder the messages here to say that update of prereq failed with err, and therefore it's content is missing?

Reply:
Hm yes, I think that would be better since it'd make the situation clearer for the user. I was trying to imply that there was a prerequisite with missing content and that we tried but failed to update it but that that's not very relevant for the user and might just confuse them